common/fs/src/cache/dir_path.rs

```diff
@@ -74,3 +83,183 @@ impl From<DirPathBuf> for PathBuf {
         d.inner
     }
 }
+
+/// If a path does not exist, the function continue to move to the parent directory
+/// until it finds a valid directory.
+///
+/// When a valid directory is found, it adds the missing piece of the directory
+/// path to `postfix` and returns the `DirPathBuf` result.
+fn find_valid_path(
+    path: Option<PathBuf>,
+    postfix: Option<PathBuf>,
+) -> Result<DirPathBuf, DirPathBufError> {
+    match path {
+        Some(p) => {
+            if matches!(std::fs::canonicalize(&p), Ok(_)) {
+                Ok(DirPathBuf { inner: p, postfix })
+            } else {
+                warn!("{:?} is not a directory; moving to parent directory", p);
+                let root_pathbuf = Path::new("/");
+                let mut postfix_pathbuf = PathBuf::new();
```

Comment:
I feel like there might be some way to use the std::path `Component` types to implement this, but that might just complicate the logic

Reply:
There's probably a few different ways to go about it. The tricky part is flipping the path each integration to make sure it comes out correct, and plays nicely with the path rules.